File: internal/capture/store.go
package capture

import (
	"database/sql"
	"strings"
)

// Schema lives in the same SQLite file as the vault index but is conceptually
// separate: the index is a cache derived from markdown, this is primary data
// that exists nowhere else. That is why retention is a deliberate setting
// rather than a cleanup detail.
const Schema = `
CREATE TABLE IF NOT EXISTS events (
    id     INTEGER PRIMARY KEY,
    ts     INTEGER NOT NULL,
    kind   TEXT    NOT NULL,
    app    TEXT,
    title  TEXT,
    url    TEXT,
    path   TEXT,
    dur_s  INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS events_ts   ON events(ts);
CREATE INDEX IF NOT EXISTS events_kind ON events(kind, ts);

-- High-water marks so pulling browser history stays incremental instead of
-- rescanning the whole file every poll.
CREATE TABLE IF NOT EXISTS source_state (
    source TEXT PRIMARY KEY,
    cursor INTEGER NOT NULL
);
`

func InitStore(db *sql.DB) error {
	_, err := db.Exec(Schema)
	return err
}

func Insert(db *sql.DB, e Event) error {
	_, err := db.Exec(
		`INSERT INTO events (ts, kind, app, title, url, path, dur_s) VALUES (?,?,?,?,?,?,?)`,
		e.TS, string(e.Kind), nullable(e.App), nullable(e.Title), nullable(e.URL), nullable(e.Path), e.DurS)
	return err
}

func InsertMany(db *sql.DB, events []Event) (int, error) {
	if len(events) == 0 {
		return 0, nil
	}
	tx, err := db.Begin()
	if err != nil {
		return 0, err
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`INSERT INTO events (ts, kind, app, title, url, path, dur_s) VALUES (?,?,?,?,?,?,?)`)
	if err != nil {
		return 0, err
	}
	defer stmt.Close()

	for _, e := range events {
		if _, err := stmt.Exec(e.TS, string(e.Kind), nullable(e.App), nullable(e.Title),
			nullable(e.URL), nullable(e.Path), e.DurS); err != nil {
			return 0, err
		}
	}
	return len(events), tx.Commit()
}

func nullable(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func Cursor(db *sql.DB, source string) int64 {
	var c int64
	if err := db.QueryRow("SELECT cursor FROM source_state WHERE source = ?", source).Scan(&c); err != nil {
		return 0
	}
	return c
}

func SetCursor(db *sql.DB, source string, cursor int64) error {
	_, err := db.Exec(
		`INSERT INTO source_state (source, cursor) VALUES (?,?)
		 ON CONFLICT(source) DO UPDATE SET cursor = excluded.cursor`, source, cursor)
	return err
}

// Range returns events in a window, oldest first.
func Range(db *sql.DB, from, to int64) ([]Event, error) {
	rows, err := db.Query(
		`SELECT id, ts, kind, app, title, url, path, dur_s FROM events
		 WHERE ts >= ? AND ts < ? ORDER BY ts`, from, to)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []Event
	for rows.Next() {
		var e Event
		var kind string
		var app, title, url, path sql.NullString
		if err := rows.Scan(&e.ID, &e.TS, &kind, &app, &title, &url, &path, &e.DurS); err != nil {
			return nil, err
		}
		e.Kind = Kind(kind)
		e.App, e.Title, e.URL, e.Path = app.String, title.String, url.String, path.String
		out = append(out, e)
	}
	return out, rows.Err()
}

// Prune drops raw events past the retention window. Rollups must already have
// extracted anything worth keeping — that is the whole contract of the two-tier
// design.
func Prune(db *sql.DB, olderThanTS int64) (int, error) {
	res, err := db.Exec("DELETE FROM events WHERE ts < ?", olderThanTS)
	if err != nil {
		return 0, err
	}
	n, err := res.RowsAffected()
	return int(n), err
}

func Count(db *sql.DB) (int, error) {
	var n int
	err := db.QueryRow("SELECT COUNT(*) FROM events").Scan(&n)
	return n, err
}

// ByIDs fetches specific events, for showing the evidence behind a proposal.
// Ordered by time rather than by id so the reader sees a narrative.
func ByIDs(db *sql.DB, ids []int64) ([]Event, error) {
	if len(ids) == 0 {
		return nil, nil
	}
	placeholders := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		placeholders[i] = "?"
		args[i] = id
	}

	rows, err := db.Query(
		`SELECT id, ts, kind, app, title, url, path, dur_s FROM events
		 WHERE id IN (`+strings.Join(placeholders, ",")+`) ORDER BY ts`, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []Event
	for rows.Next() {
		var e Event
		var kind string
		var app, title, url, path sql.NullString
		if err := rows.Scan(&e.ID, &e.TS, &kind, &app, &title, &url, &path, &e.DurS); err != nil {
			return nil, err
		}
		e.Kind = Kind(kind)
		e.App, e.Title, e.URL, e.Path = app.String, title.String, url.String, path.String
		out = append(out, e)
	}
	return out, rows.Err()
}

File: crates/capture/src/store.rs
//! Persistence for the episodic tier.
//!
//! Lives in the same SQLite file as the vault index but is conceptually
//! separate: the index is a cache derived from markdown, this is primary data
//! that exists nowhere else. It is the one part of `.brain/` that cannot be
//! rebuilt, which is why retention is a deliberate setting rather than a
//! cleanup detail.

use crate::event::{Event, Kind};
use anyhow::Result;
use rusqlite::{Connection, params};

pub const SCHEMA: &str = r#"
CREATE TABLE IF NOT EXISTS events (
    id     INTEGER PRIMARY KEY,
    ts     INTEGER NOT NULL,
    kind   TEXT    NOT NULL,
    app    TEXT,
    title  TEXT,
    url    TEXT,
    path   TEXT,
    dur_s  INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS events_ts   ON events(ts);
CREATE INDEX IF NOT EXISTS events_kind ON events(kind, ts);

-- High-water marks so pulling browser history stays incremental instead of
-- rescanning the whole file every poll.
CREATE TABLE IF NOT EXISTS source_state (
    source TEXT PRIMARY KEY,
    cursor INTEGER NOT NULL
);
"#;

pub fn init(conn: &Connection) -> Result<()> {
    conn.execute_batch(SCHEMA)?;
    Ok(())
}

pub fn insert(conn: &Connection, e: &Event) -> Result<()> {
    conn.execute(
        "INSERT INTO events (ts, kind, app, title, url, path, dur_s)
         VALUES (?1,?2,?3,?4,?5,?6,?7)",
        params![e.ts, e.kind.as_str(), e.app, e.title, e.url, e.path, e.dur_s],
    )?;
    Ok(())
}

pub fn insert_many(conn: &mut Connection, events: &[Event]) -> Result<usize> {
    let tx = conn.transaction()?;
    for e in events {
        tx.execute(
            "INSERT INTO events (ts, kind, app, title, url, path, dur_s)
             VALUES (?1,?2,?3,?4,?5,?6,?7)",
            params![e.ts, e.kind.as_str(), e.app, e.title, e.url, e.path, e.dur_s],
        )?;
    }
    tx.commit()?;
    Ok(events.len())
}

pub fn cursor(conn: &Connection, source: &str) -> i64 {
    conn.query_row("SELECT cursor FROM source_state WHERE source = ?1", params![source], |r| r.get(0))
        .unwrap_or(0)
}

pub fn set_cursor(conn: &Connection, source: &str, cursor: i64) -> Result<()> {
    conn.execute(
        "INSERT INTO source_state (source, cursor) VALUES (?1,?2)
         ON CONFLICT(source) DO UPDATE SET cursor = ?2",
        params![source, cursor],
    )?;
    Ok(())
}

/// Events in a window, oldest first.
pub fn range(conn: &Connection, from: i64, to: i64) -> Result<Vec<Event>> {
    let mut stmt = conn.prepare(
        "SELECT ts, kind, app, title, url, path, dur_s FROM events
         WHERE ts >= ?1 AND ts < ?2 ORDER BY ts",
    )?;
    let rows = stmt
        .query_map(params![from, to], |r| {
            let kind: String = r.get(1)?;
            Ok(Event {
                ts: r.get(0)?,
                kind: Kind::parse(&kind).unwrap_or(Kind::Focus),
                app: r.get(2)?,
                title: r.get(3)?,
                url: r.get(4)?,
                path: r.get(5)?,
                dur_s: r.get(6)?,
            })
        })?
        .collect::<Result<Vec<_>, _>>()?;
    Ok(rows)
}

/// Drop raw events past the retention window. Rollups must already have
/// extracted anything worth keeping — that is the whole contract of the two
/// tier design.
pub fn prune(conn: &Connection, older_than_ts: i64) -> Result<usize> {
    Ok(conn.execute("DELETE FROM events WHERE ts < ?1", params![older_than_ts])?)
}

pub fn count(conn: &Connection) -> Result<i64> {
    Ok(conn.query_row("SELECT COUNT(*) FROM events", [], |r| r.get(0))?)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn db() -> Connection {
        let c = Connection::open_in_memory().unwrap();
        init(&c).unwrap();
        c
    }

    #[test]
    fn roundtrips_events_in_a_window() {
        let c = db();
        insert(&c, &Event::new(100, Kind::Focus).app("Ghostty")).unwrap();
        insert(&c, &Event::new(200, Kind::Url).url("https://example.com")).unwrap();
        insert(&c, &Event::new(900, Kind::Focus).app("Chrome")).unwrap();

        let got = range(&c, 0, 500).unwrap();
        assert_eq!(got.len(), 2);
        assert_eq!(got[0].app.as_deref(), Some("Ghostty"));
        assert_eq!(got[1].url.as_deref(), Some("https://example.com"));
    }

    #[test]
    fn cursor_defaults_to_zero_then_persists() {
        let c = db();
        assert_eq!(cursor(&c, "chrome"), 0);
        set_cursor(&c, "chrome", 42).unwrap();
        set_cursor(&c, "chrome", 99).unwrap();
        assert_eq!(cursor(&c, "chrome"), 99);
    }

    #[test]
    fn prune_drops_only_old_rows() {
        let c = db();
        insert(&c, &Event::new(10, Kind::Focus)).unwrap();
        insert(&c, &Event::new(5000, Kind::Focus)).unwrap();
        assert_eq!(prune(&c, 1000).unwrap(), 1);
        assert_eq!(count(&c).unwrap(), 1);
    }
}
